modules/appeals_api/app/controllers/concerns/appeals_api/icn_parameter_validation.rb

```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+
+module AppealsApi
+  module IcnParameterValidation
+    extend ActiveSupport::Concern
+    # NOTE: Make sure to also include the AppealsApi::OpenidAuth concern when using this concern
+
+    ICN_REGEX = /^[0-9]{10}V[0-9]{6}$/
+
+    def validate_icn_parameter!
+      if params[:icn] && !ICN_REGEX.match?(params[:icn])
+        raise(
+          Common::Exceptions::UnprocessableEntity,
+          detail: "'icn' parameter has an invalid format. Pattern: #{ICN_REGEX.inspect}"
+        )
+      end
+
+      if veteran_icn_from_token.present? && params[:icn].present? && veteran_icn_from_token != params[:icn]
+        # If both a veteran-scoped auth token and an ICN parameter are received, the ICNs must match
+        raise(Common::Exceptions::Forbidden,
+              detail: "Invalid 'icn' parameter: Veterans may access only their own records")
+      elsif veteran_icn_from_token.blank? && params[:icn].blank?
+        # If the auth token is a system or representative token, an ICN parameter is required
+        raise(Common::Exceptions::ParameterMissing, 'icn')
+      end
```

Comment:
It sounds like we're assuming here that the Veteran ICN that's in the token (if available) will always be in the correct format. That _seems_ like a fair assumption, but I also wonder if we should log a specific error if it's not in the correct format, just in case that assumption doesn't turn out to be true once this goes to production. (We wouldn't need to tell the consumer, since they have no control of this data being valid, but we could have a log emitted to provide visibility to us.) What do you think?

Reply:
That would be something we would definitely want to address if it happened - I logged an error here to let us know